azurerm/key_credentials.go

```diff
@@ -0,0 +1,214 @@
+		}
+
+		if et, err := time.Parse(time.RFC3339, m["end_date"].(string)); err == nil {
+			buf.WriteString(fmt.Sprintf("%s-", string((et).Format(time.RFC3339))))
+		}
+	}
+
+	return hashcode.String(buf.String())
+}
+
+func flattenAzureRmKeyCredential(cred *graphrbac.KeyCredential) interface{} {
+	l := make(map[string]interface{})
+	l["key_id"] = *cred.KeyID
+	l["type"] = *cred.Type
+	l["usage"] = *cred.Usage
+	l["start_date"] = string((*cred.StartDate).Format(time.RFC3339))
+	l["end_date"] = string((*cred.EndDate).Format(time.RFC3339))
+
+	return l
+}
+
+func flattenAzureRmKeyCredentials(creds *[]graphrbac.KeyCredential) []interface{} {
+	result := make([]interface{}, 0, len(*creds))
+	for _, cred := range *creds {
+		result = append(result, flattenAzureRmKeyCredential(&cred))
+	}
+	return result
+}
+
+func expandAzureRmKeyCredential(d *map[string]interface{}) (*graphrbac.KeyCredential, error) {
+	keyId := (*d)["key_id"].(string)
+	startDate := (*d)["start_date"].(string)
+	endDate := (*d)["end_date"].(string)
+	keyType := (*d)["type"].(string)
+	usage := (*d)["usage"].(string)
+	value := (*d)["value"].(string)
+
+	if keyType == "AsymmetricX509Cert" && usage == "Sign" {
+		return nil, fmt.Errorf("Usage cannot be set to %s when %s is set as type for a Key Credential", usage, keyType)
+	}
+
+	// Match against the prefix/suffix and '\n' of certificate values.
+	// The API doesn't accept them so they need to be removed.
+	pkre := regexp.MustCompile(`(-{5}.+?-{5})|(\n)`)
+	value = pkre.ReplaceAllString(value, ``)
+
+	kc := graphrbac.KeyCredential{
+		KeyID: &keyId,
+		Type:  &keyType,
+		Usage: &usage,
+		Value: &value,
+	}
+
+	if startDate != "" {
+		starttime, sterr := time.Parse(time.RFC3339, startDate)
+		if sterr != nil {
+			return nil, fmt.Errorf("Cannot parse start_date: %q", startDate)
+		}
+		stdt := date.Time{Time: starttime.Truncate(time.Second)}
+		kc.StartDate = &stdt
+	}
+
+	if endDate != "" {
+		endtime, eterr := time.Parse(time.RFC3339, endDate)
+		if eterr != nil {
+			return nil, fmt.Errorf("Cannot parse end_date: %q", endDate)
+		}
+		etdt := date.Time{Time: endtime.Truncate(time.Second)}
+		kc.EndDate = &etdt
+	}
+
+	return &kc, nil
+}
+
+func expandAzureRmKeyCredentials(d *schema.ResourceData, o *schema.Set) (*[]graphrbac.KeyCredential, error) {
+	creds := d.Get("key_credential").(*schema.Set).List()
+	keyCreds := make([]graphrbac.KeyCredential, 0, len(creds))
+
+	for _, v := range creds {
+		cfg := v.(map[string]interface{})
+		cred, err := expandAzureRmKeyCredential(&cfg)
+		if err != nil {
+			return nil, err
+		}
+
+		// Azure only allows an in-place update of the Key Credentials list.
+		// Existing keys, matched by their KeyID, must be sent back with their
+		// Value attribute set to nil. New keys need to provide a Value.
+		// By referencing the existing schema (o), we can determine which
+		// entries in the list are existing keys.
+		if o != nil && o.Contains(v) {
+			cred.Value = nil
+		}
```

Comment:
I apologize, I'm not quite sure what you mean by the question so just in-case I'll go a bit overboard in my explanation.

When we update the list, either to change properties on existing keys, to add new keys, or to delete keys, we need to send a list to Azure that contains every single entry we want in the list after everything is said and done. There are no endpoints for creation and/or deletion of individual items. The Value field for these objects is special in that once the item has been created, the field cannot be modified. This is because it contains sensitive data and is thus write-only once. It also always returns nil on reads. To update the list with existing items, we must nil the Value field to indicate we are not changing it. All other properties on the item work the normal way for read/writes.

Example: if we have a list of items A, B, and C, and I want to delete A, leave B alone, edit C, and add D, then I must send to Azure a list with items B, C, and D. Since B and C already exist, I must nil their Value field because I cannot change it after it's been created. I can still change the other properties on C, as long as Value is left nil when I send the item back to Azure. If I don't send B back, with its Value field nil, it will get deleted even though I don't need to change any other property on it.